### PR TITLE
Experimental Features

### PR DESCRIPTION
**Pull request by terrajobst**

### CODE REVIEW

File: accepted/2021/experimental-features/experimental-features.md [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+The error displays a light-bulb in Visual Studio, suggesting to turn the feature
+on, which modifies her project file as follows:
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <ExperimentalFeatures>StaticInterfaceMembers</ExperimentalFeatures>
```
Comment (jaredpar) [inline accepted/2021/experimental-features/experimental-features.md:104]: The compiler only supports a preview mode for all features. We do not have the infrastructure for previewing individual features. As discussed we tried this in the past and it's an unmaintainable mess because the testing matrix gets out of control pretty much immediately.

**Review by jaredpar — CHANGES_REQUESTED** [review]

File: accepted/2021/experimental-features/experimental-features.md (hunk 1 above)
Comment (jaredpar) [reply accepted/2021/experimental-features/experimental-features.md:104]: Having to maintain a test matrix combination for all of the different language features we preview would be a massive cost increase. It would severely limit the amount of previewing we can do compared to what we do today

Comment (davidwrighton) [reply accepted/2021/experimental-features/experimental-features.md:104]: This is an interesting contrast to the runtime level work, where we generally have multiple experiments in place, some of which last several releases, and the typical characteristic of those experiments is that they are of ok quality, but may conflict in their implementation such that we cannot enable all experiments at once. I suspect it comes down to what is the expected quality bar for experiments, and what is their purpose. We solve the experimental test matrix problem by only testing with a single experiment enabled at once, and attempts to use them in combination is quite risky.

Comment (jaredpar) [reply accepted/2021/experimental-features/experimental-features.md:104]: Our customers don't differentiate on a feature by feature basis, they just expect to use C# Next. Once they opt in they want the features and the interaction between them. This is why we test them together because that is the expected customer use case. 

The expected quality bar is quite high here. The reliability of our implementation directly impacts the quality of Visual Studio. A huge hurdle we had to cross in even doing preview features in RTM builds was demonstrating that we could maintain quality no matter how the flag was toggled or what the user typed in the editor. 

Yes the expectation for preview features in VS is less than that of RTM features. If corner cases have extra diagnostics, bad formatting, or even if we emit bad code that's all expected. Pretty much everything else though (performance, watson, non-fatal watson, etc ...) is held to virtually the same bar.

File: accepted/2021/experimental-features/experimental-features.md [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+support is sufficiently bootstrapped, but it's the most important part in the
+validation of the feature set. Practically, this means a good chunk of the
+feature will only be available halfway through a release cycle, which greatly
+reduces the time window for gathering feedback from other parts of the product
+as well as from customers using preview builds of the upcoming .NET version.
+
+Sadly, preview feedback isn't always sufficient to judge how well a feature
+works because
+
+1. Much fewer people use preview builds of .NET and
```
Comment (stephentoub) [inline accepted/2021/experimental-features/experimental-features.md:45]:
```suggestion
1. Many fewer people use preview builds of .NET and
```

File: accepted/2021/experimental-features/experimental-features.md [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+2. Many people don't use .NET previews for their production projects, thus
+   greatly reducing the breadth & depth at which previews are being exercised.
```
Comment (stephentoub) [inline accepted/2021/experimental-features/experimental-features.md:47]: It's also difficult to explore the breadth of an impact to an ecosystem, e.g. library A depending on a feature, library B depending on library A and building on what it exposed, etc.

File: accepted/2021/experimental-features/experimental-features.md [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+
+This document is exploring a model where RTM quality builds of .NET can contain
+features that are considered experimental.
+
+The goal is to ensure that customers can discover those features & have an easy
+way to play with them while simultaneously being fully aware that these features
+aren't support yet and thus might change between versions of .NET, just like
```
Comment (stephentoub) [inline accepted/2021/experimental-features/experimental-features.md:54]:
```suggestion
aren't supported yet and thus might change (or be removed entirely) between versions of .NET, just like
```

File: accepted/2021/experimental-features/experimental-features.md [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+they would in preview builds.
```
Comment (stephentoub) [inline accepted/2021/experimental-features/experimental-features.md:55]:
```suggestion
they could in preview builds.
```

File: accepted/2021/experimental-features/experimental-features.md (hunk 1 above)
Comment (stephentoub) [reply accepted/2021/experimental-features/experimental-features.md:104]: I agree with Jared and Jonathan. I also think the separate discussions here about language vs cross-cutting features are part of the issue.  The runtime has been (at least somewhat) successful in having opt-in experimental features because they in general don't require any code changes on the part of the developer, or even rebuilds: should tiered JIT be enabled, should stack allocation of objects be enabled, should PGO be enabled, should async value task methods cache, and so on... all of these things can be changed simply via environment variables.  But the kinds of features that involve public APIs and language features do demand code changes, do entail binaries distributed with the experiment baked in, at which point it's a whole different ball game.

It's perfectly reasonable for the runtime to continue shipping experiments the way it has, assuming they're isolated from anything else in the stack as they generally have been.  But for anything that's visible via language or library, I think we should have a single opt-in mode that enables using all preview features from the language or library (and whatever runtime support is necessary to make them work).  And we should control them in a single manner: I don't have a strong opinion about whether it's a command-line switch or a property in a project file or both, but we already have langpreview, and at present that seems good enough.  It's not language vs library vs runtime; it's one stack, with one preview mode.

File: accepted/2021/experimental-features/experimental-features.md [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+        public ExperimentalFeatureAttribute(string featureName);
+        public string FeatureName { get; }
+    }
+}
+```
+
+We'll ship a built-in analyzer that will report diagnostics when an API is being
+called that is marked with this attribute and the feature name is not contained
+in the project's `ExperimentalFeatures` property. The default severity is
+`error`.
```
Comment (stephentoub) [inline accepted/2021/experimental-features/experimental-features.md:273]: In line with langpreview, I think the attribute should be recognized by the compiler itself, with the compiler error'ing just as it does when you use a language feature that's not available in the targeted language version, rather than a separate analyzer.  Maybe for VB we write an analyzer as a stop gap if we don't have the time to do the work in the compiler.  I'm not sure what solution we employ for F#.

File: accepted/2021/experimental-features/experimental-features.md (hunk 1 above)
Comment (En3Tho) [reply accepted/2021/experimental-features/experimental-features.md:104]: Would it make sense to enable another level of preview called "experimental" - the way I personally see the problem is that we, the generall mass, cannot really play with anything that is not included in preview. I mean all new fancy features are only here for us if they *actually get confirmed for the C# vnext* and thus added to preview. Experimental langversion could help with shipping features, that, for example may not require runtime support like ParameterNullChecking or fancy Attributes, Discriminated Unions and so on. Because honestly - coding those features is in no way as hard as maintaining consistency of a language and not letting it go awry. In no way "experimental" would mean that these features will get into preview and maybe C# team will have a hard time deciding that "feature X" is not for C# after all even when some people like it. I believe many features could be added to such "experimental" version of a language and people could bring real, valuable partical experience on top of existing long discussions. "Experimental" features could then be updated with new releases of .Net SDK. It seems logical because it's a bleeding edge stuff anyway.

Comment (marek-safar) [reply accepted/2021/experimental-features/experimental-features.md:104]: > The runtime has been (at least somewhat) successful in having opt-in experimental features

I think these were mostly enhancement and not full-blown features. I don't want to nitpick on naming but using different names might help here.

Otherwise, I fully agree with Stephen that we can either work on 1 feature at a time or we treat all features as one experiment.

File: accepted/2021/experimental-features/experimental-features.md [hunk 7]
```diff
@@ -0,0 +1,380 @@
+# Experimental Features
```
Comment (marek-safar) [inline accepted/2021/experimental-features/experimental-features.md:1]: I think the experimental features wording is not very encouraging for broader adoption. Our previews engagement is relative low and I think if we talk about committed features as experimental we will experience even lower engagement.

File: accepted/2021/experimental-features/experimental-features.md [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+attributes.
+
+***Note:*** F# doesn't have this capability today.
+
+### API Analyzer
+
+We'll add an attribute that we can put on APIs we consider experimental. In some
+cases it might not be possible to mark an API (for example, because it's a new
+type system feature) in which case the feature would either be guarded by new
+language syntax or by a custom analyzer.
```
Comment (jaredpar) [inline accepted/2021/experimental-features/experimental-features.md:247]: The current design doesn't seem to protect against cases like this. Let's assume for a minute that the runtime team has done the work to enable `ExperimentalFeatures` for static interfaces then someone defines the following API in the system. 

```c#
public void SubtleMethod<T>(T t) where T : ISomeName { 
  T.SomeMethod();
}
```

It's easy at a glance to miss that this is in fact using static interface members. Yet nothing I see in this doc would offer a back stop to enforce that this API was indeed marked with `ExperimentalFeature`. We're essentially depending on the diligence of code reviews. 

Yes on the consumption side the compiler would not let you consume `SubtleMethod` unless you enabled experimental features (compiler would catch the static interface usage). But at the same time nothing would stop the runtime from shipping an API with RTM expectations (lack of `ExperimentalFeature` implies it's a real API) here. 

What happens when we rip this out in later releases?

File: accepted/2021/experimental-features/experimental-features.md [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+as well as from customers using preview builds of the upcoming .NET version.
+
+Sadly, preview feedback isn't always sufficient to judge how well a feature
+works because
+
+1. Much fewer people use preview builds of .NET and
+2. Many people don't use .NET previews for their production projects, thus
+   greatly reducing the breadth & depth at which previews are being exercised.
+
+This document is exploring a model where RTM quality builds of .NET can contain
```
Comment (danmoseley) [inline accepted/2021/experimental-features/experimental-features.md:49]:
```suggestion
This document is exploring a model where GA quality builds of .NET can contain
```
Let's avoid the RTM term for the sake of consistent concepts/terms

File: accepted/2021/experimental-features/experimental-features.md [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+   that's separate)
+
+If someone shipped a binary to nuget.org that depends on an experimental
+feature, consuming it from a .NET version where the feature was release will
+still generate an error unless the consuming project has the feature turned on.
+If we believe this penalizes early adopters too much, we could make sure that
+the .NET SDK knows about which experimental features such that diagnostics for
+now released features can be suppressed.
+
+### Can we make breaking changes in experimental features?
```
Comment (danmoseley) [inline accepted/2021/experimental-features/experimental-features.md:363]: It's one thing if experiments come and go through the year. But as a platform, if the experiment goes out in an LTS release, a feature might be "experimental" for several years. Some folks may widely deploy apps that use it. Years later, it's time to upgrade and they are blocked because the experimental feature is gone. Yes, I explicitly set the experimental flag, but if the impact is large enough, it can still be hard to pull the API - just ask Windows how often they've been forced to support undocumented API or behaviors.

How can we avoid this "playing chicken" situation? The fact that once you have it in LTS, its available for several years seems to exacerbate it.

Comment (stephentoub) [reply accepted/2021/experimental-features/experimental-features.md:363]: >  just ask Windows how often they've been forced to support undocumented API or behaviors.

But Windows doesn't have side-by-side and app-local deployment models 😛

Comment (jkotas) [reply accepted/2021/experimental-features/experimental-features.md:363]: > Some folks may widely deploy apps that use it

I do not think we can sign up for supporting using experimental features in production. We need to be very clear about that.

If somebody wants to use the experimental features in production, they can only do it on their own risk.

Comment (danmoseley) [reply accepted/2021/experimental-features/experimental-features.md:363]: I agree - we should be clear that not only does experimental mean “may break or go away” but also “unsupported in production”.

I just wonder whether there is some way we can make it technically hard to use in production, to avoid the hazard on both sides.

Comment (jaredpar) [reply accepted/2021/experimental-features/experimental-features.md:363]: In our experiences on the compiler we've been able to successfully avoid people falling into this trap by 

1. Being very up front about expectations with the customers. The fact that we call it `<LangVersion>preview</LangVersion>` helps set expectations right from the start. The designs we're discussing here, using `Experimental`, is an even stronger message. 
2. Be aggressive about putting a feature per release into our preview train that isn't fully fleshed out. In part so we can get customer feedback on the feature to help shape it. In other part though to somewhat guarantee that we're going to have "break it" a couple of times during the preview. That regular break pattern helps remind customers that breaks **will** happen in preview mode. If you're a customer who doesn't that then preview is not the experience you're looking for. 
3. Stick to our principals here. We've made a few breaks that didn't go over well with customers but we pointed to the preview documentation and stuck to our design decisions.  

One advantage the compiler team has though is we ship roughly every two months in RTM releases. So we can iterate decently quickly, breaks happen in a relatively short time frame and it keeps customers from getting lulled into a false sense of security. The lead time for LTS won't have this and it may be worth thinking other ways we can help periodically remind customers about the bet they've made.

Comment (danmoseley) [reply accepted/2021/experimental-features/experimental-features.md:363]: Thought experiment. Would we have any qualms about putting out a servicing release that broke an experimental feature (intentionally or unintentionally)?

File: accepted/2021/experimental-features/experimental-features.md (hunk 7 above)
Comment (danmoseley) [reply accepted/2021/experimental-features/experimental-features.md:1]: Do you have an alternative suggestion? I don't think we should overload the 'preview' term.

In my mind experimental should be fairly well understood, quite a few apps and sites I use have "experimental" features.

File: accepted/2021/experimental-features/experimental-features.md (hunk 10 above)
Comment (jkotas) [reply accepted/2021/experimental-features/experimental-features.md:363]: > I just wonder whether there is some way we can make it technically hard to use in production, to avoid the hazard on both sides.

The plan we have been on in the past is to have the experimental features enabled for preview builds only, and disable the experimental features for the final release. It made it impossible to use the experimental features by accident in production. It would require running preview build in production by accident - if somebody does that, they have bigger problems.

This plan does not allow playing with experimental features unless you have a preview. If we go with the premise that the feature iterations really happens in previews, it does not sound too bad.

Comment (danmoseley) [reply accepted/2021/experimental-features/experimental-features.md:363]: > This plan does not allow playing with experimental features unless you have a preview. 

I cannot see that in the document -- @terrajobst if that matches your intent, could you clarify? If I understand correctly, just before GA we would push a final change that would disable the opt-in flag for experiments. That would address my concerns.

Comment (jkotas) [reply accepted/2021/experimental-features/experimental-features.md:363]: > features enabled for preview builds only, and disable the experimental features for the final release.

To be honest, this plan was not without problems. The boundary between experimental and non-experimental was fuzzy. It is what lead us to creating dotnet/runtimelab repo: https://github.com/dotnet/runtime/issues/35609 . 

I assume that we will still want to keep the more speculative experiments in dotnet/runtimelab even with this proposal. This proposal is for features that we are almost certain to ship, but we are just not able to bake them in one release cycle.

Comment (danmoseley) [reply accepted/2021/experimental-features/experimental-features.md:363]: I expect runtimelab will also remain the preferred place for experiments that require invasive changes in the shipping code paths. A good fit for this proposal would be some new API that leaves shipping code paths essentially unaffected, or a flag that forces a condition or two.

File: accepted/2021/experimental-features/experimental-features.md (hunk 7 above)
Comment (omajid) [reply accepted/2021/experimental-features/experimental-features.md:1]: FWIW, Java calls them [Preview Features])https://openjdk.java.net/jeps/12) but I agree that overloading Preview is a bad idea for us.

File: accepted/2021/experimental-features/experimental-features.md [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+
+Sadly, preview feedback isn't always sufficient to judge how well a feature
+works because
+
+1. Much fewer people use preview builds of .NET and
+2. Many people don't use .NET previews for their production projects, thus
+   greatly reducing the breadth & depth at which previews are being exercised.
+
+This document is exploring a model where RTM quality builds of .NET can contain
+features that are considered experimental.
```
[the remark this answers is not in the document]
Comment (omajid) [reply accepted/2021/experimental-features/experimental-features.md:50]: My (naive) understanding is that future RTM releases are free to (incompatibly) modify the features or even to remove them, if it really turns the feature was a really bad idea. Is that right?

File: accepted/2021/experimental-features/experimental-features.md (hunk 7 above)
Comment (drewnoakes) [reply accepted/2021/experimental-features/experimental-features.md:1]: "Experimental" suggests (to me at least) that the feature may not ever ship. "Preview" infers more commitment that the feature will one day be GA.

If the goal is to increase adoption of these features, I suggest the word implies as strong a commitment as possible. As someone evaluating a feature, the sense that this is actually going to be available in the product in future increases the chance that I try it out, as it reduces my perceived risk that it will be wholesale unavailable beyond that point.

The terms "incubation" or "pre-release" come to mind.

File: accepted/2021/experimental-features/experimental-features.md (hunk 1 above)
Comment (terrajobst) [reply accepted/2021/experimental-features/experimental-features.md:104]: My concern with having a single preview space for all features is that we won't be able to track experimental interdependencies among our own bits.

Let's say the ASP.NET team decides to ship some preview APIs for Blazor. Say the work spans multiple assemblies. The top level assembly now has to be marked as "I need preview features", which is a good thing. But it also means that the implementer of this API might accidentally end up depending on some unrelated preview BCL API or language feature. Now let's say the team promised to ship a stable version in the next release, but the BCL/language feature isn't ready for that yet. Since we know communication across larger teams is hard, I'd say it's pretty plausible that we might find this out late, putting the productization of the Blazor APIs at risk.

The same is also true for customers. Just because I want to use some experimental EF features doesn't necessarily mean I'm OK with accidentally depending on some new type system feature which I now inadvertently base my entire architecture on.

I think it's totally fine to say:

1. The compiler cannot support multiple experimental features -- it only has a preview on/off mode.
2. The runtime may or may not support multiple preview features.
3. Library layers can have multiple independent preview features (the transitivity enforcement helps to detect overlap)

We have two options to address this:

1. We give up on modeling experimental compiler features via the `ExperimentalFeature` property and ask projects to set `<LangVersion>Preview</LangVersion>` when the compiler complaints.
2. We add some custom targets to the SDK that automatically does it when specific experimental features are turned on.

But I think giving up on modeling independent experimental features for the entire stack just because certain parts of the stack cannot sensibly support independent experiments feels like throwing out the baby with the bathwater and undermine a basic philosophy of this approach: ensuring customers don't accidentally on experimental features.

Comment (jaredpar) [reply accepted/2021/experimental-features/experimental-features.md:104]: > The compiler cannot support multiple experimental features -- it only has a preview on/off mode.

I think we'll find that the compiler is not special here. I think the problems we face will likely be replicated in other tools. Consider the SDK for a moment as it's reasonable that future experiments will drift into needing SDK changes. Given their complexity lI think it's plausible they end up in the same boat here: can't afford the test matrix of a series of on / off but can't avoid it because the feature impacts VS. 

> We give up on modeling experimental compiler features via the ExperimentalFeature property and ask projects to set <LangVersion>Preview</LangVersion> when the compiler complaints.

My expectation is that if we have some tools which are on / off and some which are granular then the on / off tools flip to "preview on" whenever any of the granular features are "on". Anything else is just pushing complexity to the user for very little win.

File: accepted/2021/experimental-features/experimental-features.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (bartonjs) [reply accepted/2021/experimental-features/experimental-features.md:50]: My understanding matches Omair's, that we're talking about features that may undergo substantial breaking changes (and even potentially get cut) from one release to the next.  The incremental breaks are going to happen in preview channels (vFirst preview A to vFirst preview B, or vFirst RTM to vNext preview Alif), but the opt-in acknowledgement is for "you may end up with code that can't migrate forward or backward without a substantial rewrite".  (We change the namespace of type markers, add or remove members from the types/interfaces that power these features, etc)

File: accepted/2021/experimental-features/experimental-features.md (hunk 7 above)
Comment (marek-safar) [reply accepted/2021/experimental-features/experimental-features.md:1]: We already have a place and shipping vehicle for experiments (experimental features) and it's runtimelab. If we are going to go the extra length and invest in shipping the feature, seek the insightful developer's feedback then we cannot promote it as an experiment otherwise nobody will build anything with it.

File: accepted/2021/experimental-features/experimental-features.md (hunk 1 above)
Comment (stephentoub) [reply accepted/2021/experimental-features/experimental-features.md:104]: > But I think giving up on modeling independent experimental features for the entire stack just because certain parts of the stack cannot sensibly support independent experiments feels like throwing out the baby with the bathwater

We already have ways of shipping experiments for most runtime work, e.g. environment variables.  We already have ways of shipping experiments for most library work, e.g. as an OOB package.  I expect the majority of the things where we actually want/need to ship them as part of the actual release in a visible way (as opposed to, say, runtimelab) are where it's part of a full stack change / improvement.  Shipping numerical interfaces is useless without the language support for statics in interfaces.  Shipping UTF8 support is significantly diminished without language support for UTF8 literals (multiple folks have told us it's the most important aspect of the feature to them).  Etc.  To me, this whole thing is about the entire stack.

Comment (jkotas) [reply accepted/2021/experimental-features/experimental-features.md:104]: Roslyn has established system of feature branches, e.g. (sharplab)[https://sharplab.io/] allows you to play with them. It is not that hard to get Roslyn from feature branch combined with runtime from runtimelab, but it always going to be more hassle than just enabling a switch on top of .NET runtime that you have installed by default.

To me, this is primarily about barrier to entry to try the preview and about our own engineering efficiency (working in a branch for extended period of time is less efficient).

**Comment by terrajobst** [comment]
Converted to draft to make it clear that everything mentioned in this document is still up-in-the-air and fair game for change.

File: accepted/2021/experimental-features/experimental-features.md (hunk 10 above)
Comment (jeffhandley) [reply accepted/2021/experimental-features/experimental-features.md:363]: > Thought experiment. Would we have any qualms about putting out a servicing release that broke an experimental feature (intentionally or unintentionally)?

_I think_ it should be fair game to break experimental features (intentionally or unintentionally). I can imagine cases where we'd need to.

File: accepted/2021/experimental-features/experimental-features.md [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+## Design
+
+### Property in project files
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <ExperimentalFeatures>StaticInterfaceMembers</ExperimentalFeatures>
```
Comment (dsplaisted) [inline accepted/2021/experimental-features/experimental-features.md:192]: Shouldn't this be expressed using MSBuild items instead of properties?  For example:

```xml
<ItemGroup>
    <EnableExperimentalFeature Include="StaticInterfaceMembers" />
</ItemGroup>
```

Expressing multiple values as a semicolon-separated list for a property isn't the greatest.

File: accepted/2021/experimental-features/experimental-features.md [hunk 13]
```diff
@@ -0,0 +1,380 @@
...
+  whether they are in a runtime, a compiler, a library, or span multiple parts
+  of the system.
+* We don't have to do unnatural acts to add an experimental feature, by, for
+  example, having to build a plug-in based system that allows swapping in and
+  out files.
+
+### Non-Goals
+
+* Allowing Microsoft to turn features on or off remotely, e.g. A/B testing
+* Designing a mechanism for shipping experimental features as separate packages
```
Comment (AaronRobinsonMSFT) [inline accepted/2021/experimental-features/experimental-features.md:171]: It would be helpful to understand how this works with https://github.com/dotnet/runtimelab.

File: accepted/2021/experimental-features/experimental-features.md [hunk 14]
```diff
@@ -0,0 +1,380 @@
...
+  <PropertyGroup>
+    <ExperimentalFeatures>StaticInterfaceMembers;FancyGC</ExperimentalFeatures>
+  </PropertyGroup>
+
+</Project>
+```
+
+Would result in these assembly-level attributes:
+
+```C#
```
[the remark this answers is not in the document]
Comment (terrajobst) [reply accepted/2021/experimental-features/experimental-features.md:218]: > I would think the TFM is more important to encode than the actual feature name.

The TFM is already encoded as an assembly-level attribute today.

File: accepted/2021/experimental-features/experimental-features.md [hunk 15]
```diff
@@ -0,0 +1,380 @@
...
+  or installers
+
+## Stakeholders and Reviewers
+
+* Runtime teams
+* C#, VB, F# teams
+* Library teams
+* SDK team
+* MSBuild team
+* NuGet team
```
[the remark this answers is not in the document]
Comment (terrajobst) [reply accepted/2021/experimental-features/experimental-features.md:182]: > As a user, I want to play with preview features in VS stable. Maintaining two VS instances is too annoying, especially when there's installer bug. Instance picking also has wide impact. I've just uninstalled my VS preview yesterday. Installing a preview SDK is much easier.

This would be desirable, but is a bit outside the scope of this work. This is more about how the .NET stack itself delivers experimental features.

File: accepted/2021/experimental-features/experimental-features.md (hunk 12 above)
Comment (terrajobst) [reply accepted/2021/experimental-features/experimental-features.md]: Fixed

File: accepted/2021/experimental-features/experimental-features.md (hunk 8 above)
Comment (terrajobst) [reply accepted/2021/experimental-features/experimental-features.md:258]: It's called out in the section you're standing on:

> We'll add an attribute that we can put on APIs we consider experimental. In some cases it might not be possible to mark an API (for example, because it's a new type system feature) in which case the feature would either be guarded by new language syntax or **by a custom analyzer**.

But I would generally assume that if we apply the attribute to an interface, we'd flag implementations and and interface constraints as well.

Comment (jaredpar) [reply accepted/2021/experimental-features/experimental-features.md:258]: I think you're missing my point here. My point isn't that we lack mechanisms to enforce this pattern when we see it. My point is there are no mechanisms in place to enforce we add these mechanisms where needed. Basically any one could come along, put this into a PR and unless a human was there to spot it lacked the attribute nothing else would stop it from being merged. Then we'd be left with an experimental API that wasn't marked as such.

File: accepted/2021/experimental-features/experimental-features.md [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+already has a mechanism for this which is exposed to both the compiler as well
+as for analyzers/source generators. Alternatively, we could say the information
+about experimental features can be retrieved from the generated assembly-level
+attributes.
+
+***Note:*** F# doesn't have this capability today.
+
+### API Analyzer
+
+We'll add an attribute that we can put on APIs we consider experimental. In some
```
Comment (jkotas) [inline accepted/2021/experimental-features/experimental-features.md:255]: This will protect compile time. Do we also need to protect runtime time, like reflection exposing the experimental surface, and apps and libraries taking accidental dependency on it?

File: accepted/2021/experimental-features/experimental-features.md [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+help to speed things up considerably. They decide to give it a spin, so they
+copy & paste some sample code from the blog post.
+
+The code immediately produces an error message:
+
+> error: The type 'IArithmetic\<T>' is part of the experimental feature
+> 'StaticInterfaceMembers' which must be turned on explicitly.
+
+After taking a quick look at the blog post, Ainsley realizes that they only need
+to add an `ExperimentalFeature` item to the project files.
```
Comment (jcouv) [inline accepted/2021/experimental-features/experimental-features.md:136]: nit: we should be able to offer the lightbulb onboarding experience here too.

File: accepted/2021/experimental-features/experimental-features.md [hunk 18]
```diff
@@ -0,0 +1,390 @@
...
+this issue. Adriana does a quick web search and finds a blog post from the C#
+team that showcases how this feature might look like. She wonders how she can
+plan with that. Optimistic, she copy & pastes the sample code into her project
+which immediately generates a compilation error:
+
+> error: static interface members are an experimental feature of C# 9.0 and must
+> be turned on explicitly.
+
+The error displays a light-bulb in Visual Studio, suggesting to turn the feature
+on, which modifies her project file as follows:
```
Comment (jcouv) [inline accepted/2021/experimental-features/experimental-features.md:100]: Since we offer such lightbulb experience already (UpgradeLangVer), we'll have to decide whether to offer just the new one or both.
It feels like offering just the new one should be enough.

File: accepted/2021/experimental-features/experimental-features.md [hunk 19]
```diff
@@ -0,0 +1,390 @@
...
+                    AttributeTargets.Struct |
+                    AttributeTargets.Enum |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Method |
+                    AttributeTargets.Property |
+                    AttributeTargets.Field |
+                    AttributeTargets.Event, AllowMultiple = true, Inherited = false)]
+    public sealed class ExperimentalFeatureAttribute : Attribute
+    {
+        public ExperimentalFeatureAttribute(string featureName);
```
Comment (jcouv) [inline accepted/2021/experimental-features/experimental-features.md:275]: nit: assuming we keep multiple feature flags, as opposed to a single "preview" flag, we may want `params string[] featureNames` instead for compactness. Then we may reconsider `AllowMultiple` option.

File: accepted/2021/experimental-features/experimental-features.md [hunk 20]
```diff
@@ -0,0 +1,390 @@
...
+
+### API Analyzer
+
+We'll add an attribute that we can put on APIs we consider experimental. In some
+cases it might not be possible to mark an API (for example, because it's a new
+type system feature) in which case the feature would either be guarded by new
+language syntax or by a custom analyzer.
+
+```C#
+namespace System.Runtime.Versioning
```
Comment (jcouv) [inline accepted/2021/experimental-features/experimental-features.md:261]: nit: consider re-using namespace where we have [RuntimeFeature](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.runtimefeature?view=net-5.0) already.

File: accepted/2021/experimental-features/experimental-features.md [hunk 21]
```diff
@@ -0,0 +1,390 @@
...
+                    AttributeTargets.Event, AllowMultiple = true, Inherited = false)]
+    public sealed class ExperimentalFeatureAttribute : Attribute
+    {
+        public ExperimentalFeatureAttribute(string featureName);
+        public string FeatureName { get; }
+    }
+}
+```
+
+We'll ship a built-in analyzer that will report diagnostics when an API is being
```
Comment (jcouv) [inline accepted/2021/experimental-features/experimental-features.md:281]: nit: not sure if it'll be an analyzer, the compiler(s) themselves, or a combination. I'd rephrase as "We will report diagnostics ..."

File: accepted/2021/experimental-features/experimental-features.md (hunk 10 above)
Comment (terrajobst) [reply accepted/2021/experimental-features/experimental-features.md:373]: I have added section about quality bar that should address this point.

File: accepted/2021/experimental-features/experimental-features.md [hunk 22]
```diff
@@ -0,0 +1,390 @@
...
+called that is marked with this attribute and the feature name is not contained
+in the project's `ExperimentalFeature` items. The default severity is `error`.
+
+This analyzer will also validate that all referenced assemblies only have
```
Comment (jcouv) [inline accepted/2021/experimental-features/experimental-features.md:285]: I think the analysis should be finer-grain. The parts of the conuming project that consume parts of a reference assembly that is in "preview" should themselves be marked as in "preview".

Let's say we add `INumeric` to the BCL as "preview" feature and some upper layers of the BCL expose more types or APIs on top of it. Those specific upper parts should be marked as "preview" (no less).

File: accepted/2021/experimental-features/experimental-features.md [hunk 23]
```diff
@@ -0,0 +1,390 @@
...
+
+would result in a diagnostic like this:
+
+> error: Assembly 'Foo.Bar' depends on the experimental feature 'FancyGC' which
+> the consuming project has not enabled.
+
+***Note:*** Since F# doesn't support analyzers, we would have to add this
+capability to the compiler itself.
+
+## Q & A
```
Comment (jcouv) [inline accepted/2021/experimental-features/experimental-features.md:314]: Should the Intellisense and the debugger show all types/APIs, or filter based on the current preview features that are enabled?
To increase adoption of preview features, we may not want to filter the APIs. But we may want to display them differently, to avoid annoyance (I keep taking a dependency that I'm not allowed to).

File: accepted/2021/experimental-features/experimental-features.md (hunk 16 above)
Comment (jkotas) [reply accepted/2021/experimental-features/experimental-features.md:255]: This includes indirect dependencies on reflection, like Powershell. Should Powershell be invited to this conversation to provide feedback? It may require work on their side.

File: accepted/2021/experimental-features/experimental-features.md (hunk 10 above)
Comment (danmoseley) [reply accepted/2021/experimental-features/experimental-features.md:373]: OK @jkotas note that this is not the plan:

> This plan does not allow playing with experimental features unless you have a preview.

Comment (jkotas) [reply accepted/2021/experimental-features/experimental-features.md:373]: In practice, I expect you will want to have a preview installed to play with experimental features. Consider this:

1. We ship a .NET release, the VS/compiler sync-shipping with  the .NET release will work well. 
2. One or two months later, we ship first vNext preview and VS will pick up matching changes. It is safe to assume that the vNext will include some breaking changes in the preview features. It means that the current VS is not going to work great with current stable release. For example, you may get a ton of red squiggles everywhere.
3. It is between hard and impossible to downgrade VS. Your most viable options is going to be to target vNext preview runtimes; or lose the VS bells and whistles.

We need to optimize this on getting feedback and iterating in preview .NET release. @jaredpar made the same point in https://github.com/dotnet/designs/pull/185/#discussion_r590483029 . Latest stable runtime + current VS is likely going to be poor experience most of the time.

With this point of view, I think it would be ok to drop the preview features from the official runtime release. It would be good hygiene and address all issue with accidental dependencies.

Comment (jkotas) [reply accepted/2021/experimental-features/experimental-features.md:373]: > We need to optimize this on getting feedback and iterating in preview .NET releases

This also helps with focus and marketing. When we ship a new release, the message coming with it is "go upgrade your apps now". We may not want to dilute it with asking for feedback about preview features at the same time.

**Comment by terrajobst** [comment]
Superseded by #192.
